### DISCUSSION TITLE
make error:  error: ‘missing_errfile_declaration’ is deprecated [-Werror=deprecated-declarations]

### DISCUSSION THREAD

**Discussion by xuchao1234** [ipxe/ipxe#900] [Q&A] [answered]
I tried to support ROM compilation of a new network card based on the virtnet driver, so I copied the virtio-net.c and virtio-net.h files and modified them to fit my virtio-net network card, but the compilation reported an error. As long as I use the macros similar to ENOENT and EINVAL defined in ipxe-master/src/include/errno.h, an error will be reported. How can I solve this problem?

![image](https://user-images.githubusercontent.com/88670337/220867134-485e0434-4091-40df-ab84-21f667e5358a.png)

**Comment by mcb30** [comment, marked as answer]
As part of your pull request to add the new driver, you would need to include entries in `include/ipxe/errfile.h` that assign a unique error file number for your new driver's file.  If the pull request is merged, then the numbers would be officially allocated at the point of merging.

The version of `include/ipxe/errfile.h` in the iPXE master branch acts as the official list of allocated error file numbers, and you may end up with a different allocated number than whatever you choose to use during development so you should not distribute any binaries before getting this official allocation.

What is different about your version of the `virtio-net` driver?  It seems quite unlikely that whatever you are trying to do actually requires a separate driver file.

**Reply by xuchao1234** [reply]
Thank you! We are developing a new network card, currently based on FPGA verification, and this network card is virtio-net. In order to verify the PXE functionality, I need to modify it based on the virtio-net driver. When distributing, I will do what you suggested.

**Reply by mcb30** [reply]
Interesting.  `virtio` is designed specifically to optimise guest-to-host communication, and is not well suited for a real hardware device.  Why did you choose to use the `virtio` abstraction as a basis for your hardware?

**Reply by xuchao1234** [reply]
We only use the virtio-net queues to send and receive packets, and we will redesign the hardware to support other functions and verify them one by one. Finally, we will completely replace all the features of virtio and redesign them.
